mesonbuild/compilers/c.py

```diff
@@ -228,9 +228,10 @@ def get_default_include_dirs(self):
         return []
 
     def gen_export_dynamic_link_args(self, env):
-        if for_windows(env.is_cross_build(), env) or for_cygwin(env.is_cross_build(), env):
+        m = env.machines[self.for_machine]
+        if m.is_windows(env) or m.is_cygwin():
```

Comment:
Code like `for_windows(env.is_cross_build(), env)` is very suspicious because it breaks on / precludes `native: true` dependencies. `env.machines[self.for_machine],for_windows()` tries to force the programmer to pull the machine from the target at hand, and works the same way whether in native or cross builds.